### PR TITLE
[ROCm] Cache fp32 upcast of static e8m0 weight scale in AITER scaled_mm

### PR DESCRIPTION
**Pull request by jiacao-amd**
## Purpose

The e8m0 weight scale in the AITER block-scaled GEMM path is static, but
`apply_block_scaled_mm` re-runs the `<<23` bit-shift fp32 upcast plus
`.contiguous()` on it **every decode step, for every layer**, even though the
value never changes.

On DeepSeek-V4 FP4 (MI355X / gfx950) this shows up in profiles as ~3.5% of GPU
time split across `aten::__lshift__` and `direct_copy`.

This PR overrides `process_weights_after_loading` in `AiterFp8BlockScaledMMKernel`
to convert the e8m0 weight scale to fp32 once at model load time via
`replace_parameter`. The per-step activation scale (`As`) is dynamic and remains
handled in `apply_block_scaled_mm` unchanged.

## Test Plan

- DeepSeek-V4-Pro FP4, TP8, MI355X, 1024/1024 decode sweep (conc 16–128).
- gsm8k 5-shot (full 1319 samples) accuracy parity check.
- Verify `__lshift__` / `direct_copy` ops disappear from the decode profile.

### Throughput / latency (8× MI355X, DeepSeek-V4-Pro FP4)

8× MI355X (gfx950), TP=8 · DeepSeek-V4-Pro (FP4 MoE + FP8 attention), KV-cache fp8 · default fused AITER MoE · random ISL/OSL 1024/1024, num-prompts = 10×concurrency. Off → on = same base without / with this patch.

**Performance**

| concurrency | output tok/s (off → on) | Δ | mean TPOT ms (off → on) | Δ |
|---|---|---|---|---|
| 16  | 542.48 → 590.22  | +8.8% | 28.10 → 26.02 | −7.4% |
| 32  | 937.39 → 1006.73 | +7.4% | 34.06 → 30.74 | −9.7% |
| 64  | 1504.27 → 1591.14 | +5.8% | 42.30 → 39.10 | −7.6% |
| 128 | 2295.98 → 2392.37 | +4.2% | 55.92 → 53.22 | −4.8% |

Note: median TTFT at c16 is dominated by request-queue scheduling noise (±38% run-to-run with no code change, confirmed via 3-pass repeatability on a warm server); TPOT and throughput are stable across repetitions.

**Accuracy — gsm8k 5-shot, full 1319 samples (exact_match)**

| | off | on |
|---|---|---|
| strict-match     | 0.9500 | 0.9484 |
| flexible-extract | 0.9500 | 0.9477 |

### CODE REVIEW

**Comment by Fangzhou-Ai** [comment]
Hi @tjtanaa  @dllehr-amd  can you take a look at this PR?

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py [hunk 1]
```diff
@@ -375,6 +375,24 @@ def __init__(self, config: FP8ScaledMMLinearLayerConfig):
             and rocm_aiter_ops.is_triton_gemm_w8a8_tuned(n, k)
         )
 
+    def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        from vllm.model_executor.layers.quantization.utils.fp8_utils import (
+            _upcast_e8m0_to_fp32,
+        )
+        from vllm.model_executor.utils import replace_parameter
```
Comment (dllehr-amd) [inline vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: should we need another import here?  isn't it imported at the top?

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py [hunk 2]
```diff
@@ -375,6 +375,24 @@ def __init__(self, config: FP8ScaledMMLinearLayerConfig):
...
+
+        from .BlockScaledMMLinearKernel import FP8BlockParams
```
Comment (dllehr-amd) [inline vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: In the same vein, there's already an import at the top for BlockScaledMMLinearKernel  functions, can you add it here?  i understand you're trying to lazy import on this function, but it'll be cleaner to just add it above

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py
Comment (dllehr-amd) [inline vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: we are now importing this twice, may be time to move it to the top as well.

**Review by dllehr-amd — CHANGES_REQUESTED** [review]
Hey @jiacao-amd  Nice catch!  I'm mostly good with the work, just a small cleanup with the imports.  As an aside, have you looked at other scale shapes besides the e8m0fnu?  I know the current impl didn't, but for uint8 or others we don't do anything but send it to fp32.

**Comment by jiacao-amd** [comment]
@dllehr-amd Good question, I traced the quant path that can select this kernel, and the block weight scale is only registered as float8_e8m0fnu or float32. So upcast e8m0 here should be enough.

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py (hunk 1 above)
Comment (jiacao-amd) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: done, move to top

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py (hunk 2 above)
Comment (jiacao-amd) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: done

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py
Comment (jiacao-amd) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: removed

**Review by dllehr-amd — APPROVED** [review]
Thanks @jiacao-amd   This looks good to me!
